### PR TITLE
GH-35506: [C++] Support First and Last aggregators in Substrait

### LINKED ISSUE #35506: [C++] Support First and Last aggregators in Substrait
### Describe the enhancement requested

Currently, Ibis-Substrait defines First and Last aggregators while Arrow-Substrait does not support them. This issue will add the support.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by rtpsw**
### Rationale for this change

See #35506

### What changes are included in this PR?

"first" and "last" are registered as aggregates in Substrait.

### Are these changes tested?

Yes, The test-cases cover grouped and non-grouped aggregation with "first" and "last", as well as "null" and non-"null" outputs.

### Are there any user-facing changes?

Yes, "first" and "last" will be valid in a consumed Substrait plan.
* Closes: #35506

### CODE REVIEW

**Comment by rtpsw** [comment]
cc @icexelloss

**Comment by rtpsw** [comment]
cc @icexelloss

**Comment by rtpsw** [comment]
This [CI job](https://github.com/apache/arrow/actions/runs/4926654580/jobs/8802535799?pr=35513) has an unrelated error that seems worth looking into separately.

File: cpp/src/arrow/engine/substrait/function_test.cc [hunk 1]
```diff
@@ -574,6 +575,8 @@ struct AggregateTestCase {
   std::shared_ptr<DataType> output_type;
   // The aggregation takes zero columns as input
   bool nullary = false;
+  // The aggregation is ordered
+  bool ordered = false;
```
Comment (icexelloss) [inline cpp/src/arrow/engine/substrait/function_test.cc:579]: If aggregation is ordered, shouldn't this be true?

File: cpp/src/arrow/engine/substrait/serde.h [hunk 2]
```diff
@@ -96,6 +96,26 @@ ARROW_ENGINE_EXPORT Result<std::shared_ptr<acero::ExecPlan>> DeserializePlan(
...
+/// \param[in,out] plan an ExecNode corresponding to the single toplevel relation in the
+/// Substrait Plan
+/// \param[in] buf a buffer containing the protobuf serialization of a Substrait Plan
+/// message
+/// \param[in] consumer node that consumes the batches produced by each toplevel Substrait
+/// relation
+/// \param[in] registry an extension-id-registry to use, or null for the default one.
+/// \param[out] ext_set_out if non-null, the extension mapping used by the Substrait
+/// \param[in] conversion_options options to control how the conversion is to be done.
+ARROW_ENGINE_EXPORT Status DeserializePlan(
```
Comment (icexelloss) [inline cpp/src/arrow/engine/substrait/serde.h:113]: Why this change

File: cpp/src/arrow/engine/substrait/serde.h [hunk 3]
```diff
@@ -133,13 +153,34 @@ ARROW_ENGINE_EXPORT Result<std::vector<acero::Declaration>> DeserializePlans(
...
+/// Substrait Plan
+/// \param[in] buf a buffer containing the protobuf serialization of a Substrait Plan
+/// message
+/// \param[in] write_options write options of a node consuming the batches produced by
+/// each toplevel Substrait relation
+/// \param[in] registry an extension-id-registry to use, or null for the default one.
+/// \param[out] ext_set_out if non-null, the extension mapping used by the Substrait
+/// Plan is returned here.
+/// \param[in] conversion_options options to control how the conversion is to be done.
+ARROW_ENGINE_EXPORT Status DeserializePlan(
```
Comment (icexelloss) [inline cpp/src/arrow/engine/substrait/serde.h:178]: Why this change?

File: cpp/src/arrow/engine/substrait/serde.h (hunk 2 above)
Comment (icexelloss) [reply cpp/src/arrow/engine/substrait/serde.h:113]: Also, shouldn't this return Result object similar to other Deserialize* function?

**Comment by icexelloss** [comment]
@rtpsw there seems to be some refactoring changes. It is not clear why those changes are in this PR, can you comment?

**Comment by icexelloss** [comment]
> What changes are included in this PR?
> "first" and "last" are registered as aggregates in Substrait.

There are more changes then just "register" first and last aggregates it seems. It would better if you can explain them upfront.

File: cpp/src/arrow/engine/substrait/function_test.cc (hunk 1 above)
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/function_test.cc:579]: The pre-PR behavior is equivalent to post-PR behavior with `ordered = false`. In pre-PR, all tested aggregators had no trouble with the default executor. In post-PR, the `first` and `last` aggregators are ordered and with `ordered = false` there is an error `NotImplemented: Using ordered aggregator in multiple threaded execution is not supported`.

File: cpp/src/arrow/engine/substrait/serde.cc [hunk 4]
```diff
@@ -236,14 +236,13 @@ ARROW_ENGINE_EXPORT Result<PlanInfo> DeserializePlan(
 
 namespace {
 
-Result<std::shared_ptr<acero::ExecPlan>> MakeSingleDeclarationPlan(
-    std::vector<acero::Declaration> declarations) {
+Status MakeSingleDeclarationPlan(acero::ExecPlan* plan,
```
[the remark this answers is not in the document]
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/serde.cc:239]: This allows one to pass a preconfigured plan instead of getting a default-configured plan. The tester uses this to pass a plan with a non-default executor `ctx_for_ordered`.

File: cpp/src/arrow/engine/substrait/serde.h (hunk 2 above)
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/serde.h:113]: Same rationale as in [this post](https://github.com/apache/arrow/pull/35513#discussion_r1188823725).

File: cpp/src/arrow/engine/substrait/serde.h (hunk 3 above)
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/serde.h:178]: Same rationale as in [this post](https://github.com/apache/arrow/pull/35513#discussion_r1188823725).

**Review by westonpace — COMMENTED** [review]
I don't see `first` / `last` in https://github.com/substrait-io/substrait/blob/main/extensions/functions_aggregate_generic.yaml

So wouldn't this be an Arrow extension function and we should use `kArrowSimpleExtensionFunctionsUri`.

**Comment by rtpsw** [comment]
> I don't see first / last in https://github.com/substrait-io/substrait/blob/main/extensions/functions_aggregate_generic.yaml
>
> So wouldn't this be an Arrow extension function and we should use kArrowSimpleExtensionFunctionsUri.

You're right, this PR is ahead of Ibis-Substrait. @icexelloss, should we create a corresponding PR for Ibis-Substrait?

**Comment by icexelloss** [comment]
> You're right, this PR is ahead of Ibis-Substrait. @icexelloss, should we create a corresponding PR for Ibis-Substrait?

Yeah I think so, since first and last is defined in ibis now

File: cpp/src/arrow/engine/substrait/serde.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (icexelloss) [reply cpp/src/arrow/engine/substrait/serde.cc:239]: I think you can use https://github.com/apache/arrow/blob/c3acc91c3e21f225ee7cc3cb3f489f0153365f5a/cpp/src/arrow/acero/exec_plan.h#L566

Instead and can avoid changing the serde files

[the remark this answers is not in the document]
Comment (icexelloss) [reply cpp/src/arrow/engine/substrait/serde.cc:239]: i.e., instead of de serializing into a plan, you can de serializing into a declaration and use single-threaded or multi-threaded execution depending on ordered/non-ordered.

[the remark this answers is not in the document]
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/serde.cc]: Done. It turned out not easy to use `DeclarationToTable` here because of the extra setup needed for linking the input table and the consumer, but I did separate out the deserialization and obtained simpler code.

File: cpp/src/arrow/engine/substrait/serde.h [hunk 5]
```diff
@@ -89,8 +89,7 @@ ARROW_ENGINE_EXPORT Result<std::vector<acero::Declaration>> DeserializePlans(
 /// \param[out] ext_set_out if non-null, the extension mapping used by the Substrait
 /// \param[in] conversion_options options to control how the conversion is to be done.
 /// Plan is returned here.
-/// \return an ExecNode corresponding to the single toplevel relation in the Substrait
-/// Plan
+/// \return an ExecPlan for the Substrait Plan
```
Comment (icexelloss) [inline cpp/src/arrow/engine/substrait/serde.h:92]: Revert this?

File: cpp/src/arrow/engine/substrait/serde.h [hunk 6]
```diff
@@ -133,8 +132,7 @@ ARROW_ENGINE_EXPORT Result<std::vector<acero::Declaration>> DeserializePlans(
 /// \param[out] ext_set_out if non-null, the extension mapping used by the Substrait
 /// Plan is returned here.
 /// \param[in] conversion_options options to control how the conversion is to be done.
-/// \return a vector of ExecNode declarations, one for each toplevel relation in the
-/// Substrait Plan
+/// \return an ExecPlan for the Substrait Plan
```
Comment (icexelloss) [inline cpp/src/arrow/engine/substrait/serde.h:135]: Revert this?

File: cpp/src/arrow/engine/substrait/function_test.cc (hunk 1 above)
Comment (icexelloss) [reply cpp/src/arrow/engine/substrait/function_test.cc:578]: Can you add some comment to make this more clear? Just looking at
```
  // The aggregation is ordered
  bool ordered = false;
```
It's kind of confusing

**Comment by icexelloss** [comment]
@rtpsw Left some comments

**Review by icexelloss — CHANGES_REQUESTED** [review]
Overall looks good. Left some comments.

File: cpp/src/arrow/engine/substrait/function_test.cc [hunk 7]
```diff
@@ -601,6 +603,7 @@ std::shared_ptr<Table> GetOutputTableForAggregateCase(
 std::shared_ptr<acero::ExecPlan> PlanFromAggregateCase(
     const AggregateTestCase& test_case, std::shared_ptr<Table>* output_table,
     bool with_keys) {
+  static ExecContext ctx_for_ordered;
```
[the remark this answers is not in the document]
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/function_test.cc:606]: The `static` makes the `ExecContext` instance global (and restricted to the function's scope), preventing its premature destruction. This is useful because the method it is passed to takes a pointer, rather than a `shared_ptr`, to an `ExecContext`. This instance is only used when the aggregation-case is ordered.

File: cpp/src/arrow/engine/substrait/serde.h (hunk 5 above)
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/serde.h:92]: The pre-PR comment is incorrect, so I fixed it.

File: cpp/src/arrow/engine/substrait/serde.h (hunk 6 above)
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/serde.h:135]: The pre-PR comment is incorrect, so I fixed it.

File: cpp/src/arrow/engine/substrait/function_test.cc [hunk 8]
```diff
@@ -626,10 +629,22 @@ std::shared_ptr<acero::ExecPlan> PlanFromAggregateCase(
...
+  struct TestConsumerFactory {
+    std::shared_ptr<acero::SinkNodeConsumer> operator()() { return consumer; }
+    std::shared_ptr<acero::SinkNodeConsumer> consumer;
+  };
   EXPECT_OK_AND_ASSIGN(
-      std::shared_ptr<acero::ExecPlan> plan,
-      DeserializePlan(*substrait, std::move(consumer), default_extension_id_registry(),
-                      /*ext_set_out=*/nullptr, conversion_options));
+      auto declarations,
+      DeserializePlans(*substrait, TestConsumerFactory{std::move(consumer)},
```
[the remark this answers is not in the document]
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/function_test.cc:638]: No, because the function takes a `ConsumerFactory` instance, and a lambda won't compile here.

**Comment by westonpace** [comment]
> You're right, this PR is ahead of Ibis-Substrait. @icexelloss, should we create a corresponding PR for Ibis-Substrait?

Wouldn't the PR need to be in Substrait itself?  If you want to use https://github.com/substrait-io/substrait/blob/main/extensions/functions_aggregate_generic.yaml then it would seem that file is the one that needs to change.

**Comment by westonpace** [comment]
ibis-substrait and arrow have to agree on function names (and by name I mean both the URI and the name).

substrait-io/substrait is the main repository for substrait and defines a standard set of functions that are likely to be interesting to all producers and consumers.  I do think first/last belong in that list, however, there may be some discussion to figure out exactly how to classify them (no existing aggregate functions today depend on order so are these aggregate functions or some kind of special new function).

So far, to my understanding, functions have fallen into two categories:

 * Well defined functions
 
 These functions are defined in substrait-io/substrait.  They have "official" names that are part of the Substrait spec and should be long standing.  In this case, agreement is easy, official names should always be preferred.  To use an official name it is important that Arrow has a mapping that maps the official name to the Arrow function name.

 * Arrow-specific functions

These functions are available in Arrow, but not defined yet in substrait-io/substrait.  Since there is no official name we instead use a special URI `urn:arrow:substrait_simple_extension_function` with the Arrow function name.  One current limitation of Arrow-specific functions is that we cannot specify options.  However, one benefit is that no mapping is required.  The function name in the Substrait plan should always match the Arrow function name exactly.

I'm confused by this PR.  It is adding an Arrow mapping, but there is no official function (yet).  It will "work" if both ibis-substrait & arrow act like it's already in Substrait but I feel this is going to be confusing to people.  Anyone looking at the Arrow repo, for example, might think that there really is an official first/last function.  I think it's best to use the arrow-specific URI until the official function is adopted.

**Comment by icexelloss** [comment]
Thanks @westonpace for catching this and explain in details. I agree with you.

To be clear, is the ask to use "urn:arrow:substrait_simple_extension_function" instead of "https://github.com/substrait-io/substrait/blob/main/extensions/functions_aggregate_generic.yaml" for the function URI?

File: cpp/src/arrow/engine/substrait/function_test.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (icexelloss) [reply cpp/src/arrow/engine/substrait/function_test.cc:606]: This still feel strange to me that this is needed. Does something like this not work?
https://github.com/apache/arrow/blob/401ae19051e9120fedb8372a13cd68e08798c00a/cpp/src/arrow/acero/benchmark_util.cc#L49

[the remark this answers is not in the document]
Comment (icexelloss) [reply cpp/src/arrow/engine/substrait/function_test.cc:606]: Looks like you can use this one?
https://github.com/apache/arrow/blob/main/cpp/src/arrow/acero/exec_plan.h#L72

**Comment by westonpace** [comment]
> To double check, is the ask to use "urn:arrow:substrait_simple_extension_function" instead of "https://github.com/substrait-io/substrait/blob/main/extensions/functions_aggregate_generic.yaml" for the function URI?

Yes.  Furthermore, if you use `urn:arrow:substrait_simple_extension_function` then you don't need a mapping at all.

However, since this is an aggregate function, I suspect we will need to fix `arrow::engine::ExtensionIdRegistry::SubstraitAggregateToArrow` so that it behaves similarly to `arrow::engine::ExtensionIdRegistry::SubstraitCallToArrow`:

```
  Result<SubstraitCallToArrow> GetSubstraitCallToArrow(
      Id substrait_function_id) const override {
    // NEED TO ADD THIS CASE TO GetSubstraitAggregateToArrow
    if (substrait_function_id.uri == kArrowSimpleExtensionFunctionsUri) {
      return kSimpleSubstraitToArrow;
    }
    auto maybe_converter = substrait_to_arrow_.find(substrait_function_id);
    if (maybe_converter == substrait_to_arrow_.end()) {
      if (parent_) {
        return parent_->GetSubstraitCallToArrow(substrait_function_id);
      }
      return Status::NotImplemented(
          "No conversion function exists to convert the Substrait function ",
          substrait_function_id.uri, "#", substrait_function_id.name,
          " to an Arrow call expression");
    }
    return maybe_converter->second;
  }
```

**Comment by rtpsw** [comment]
Thanks, @westonpace. I'll work on a revision accordingly.

File: cpp/src/arrow/engine/substrait/extension_set.cc [hunk 9]
```diff
@@ -1113,9 +1113,11 @@ struct DefaultExtensionIdRegistry : ExtensionIdRegistryImpl {
                                            DecodeBasicAggregate("mean")));
     DCHECK_OK(AddSubstraitAggregateToArrow({kSubstraitArithmeticFunctionsUri, "std_dev"},
                                            DecodeBasicAggregate("stddev")));
-    DCHECK_OK(
-        AddSubstraitAggregateToArrow({kSubstraitAggregateGenericFunctionsUri, "count"},
-                                     DecodeBasicAggregate("count")));
+    for (const auto& fn_name : {"count", "first", "last"}) {
+      DCHECK_OK(
+          AddSubstraitAggregateToArrow({kSubstraitAggregateGenericFunctionsUri, fn_name},
```
Comment (icexelloss) [inline cpp/src/arrow/engine/substrait/extension_set.cc:1118]: Should we uri the arrow specifc uri for first/last instead of the substrait one?

Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/extension_set.cc:1118]: Right. Fixed.

File: cpp/src/arrow/engine/substrait/function_test.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/function_test.cc]: OK, I did something similar.

File: cpp/src/arrow/engine/substrait/serde.h (hunk 6 above)
Comment (icexelloss) [reply cpp/src/arrow/engine/substrait/serde.h:135]: Oh you are right I am looking at the wrong thing

File: cpp/src/arrow/engine/substrait/function_test.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/function_test.cc:606]: Here are [a couple of instances](https://github.com/search?q=repo%3Aapache%2Farrow%20%22static%20ExecContext%22&type=code) where this is used.

[the remark this answers is not in the document]
Comment (icexelloss) [reply cpp/src/arrow/engine/substrait/function_test.cc:606]: Ok since there is existing usage of this pattern I am ok with it.

File: cpp/src/arrow/engine/substrait/function_test.cc (hunk 1 above)
Comment (rtpsw) [reply cpp/src/arrow/engine/substrait/function_test.cc:578]: Clarified.

**Review by icexelloss — APPROVED** [review]
LGTM

**Comment by icexelloss** [comment]
@westonpace I am gonna go head and merge this one. If you have any comments please let @rtpsw know and he can address as follow up.
